### PR TITLE
[posix] open netif fds with O_NONBLOCK in a compatible way

### PR DESCRIPTION
**Pull request by gjc13**
Resolves https://github.com/openthread/openthread/issues/4770

### CODE REVIEW

**Comment by rob-the-dude** [comment]
I understand that the tunnel descriptor is not a socket, but should we open it in non-blocking mode as well?  It's opened in platformNetifInit -- I think you could add " | O_NONBLOCK" to the open call to get the desired behavior.

**Comment by gjc13** [comment]
> I understand that the tunnel descriptor is not a socket, but should we open it in non-blocking mode as well? It's opened in platformNetifInit -- I think you could add " | O_NONBLOCK" to the open call to get the desired behavior.

Yes, see https://stackoverflow.com/questions/5351994/will-read-ever-block-after-select for why we need to specify O_NONBLOCK even if we are using it with select.

**Comment by gjc13** [comment]
@bukepo Looks like virtual time doesn't work well with O_NONBLOCK so I still need to leave it in blocking mode.

File: src/posix/platform/misc.cpp [hunk 1]
```diff
@@ -107,6 +107,31 @@ int SocketWithCloseExec(int aDomain, int aType, int aProtocol)
...
+
+    VerifyOrExit((rval = fcntl(fd, F_GETFD, 0)) != -1, perror("fcntl(F_GETFD)"));
+    VerifyOrExit((rval = fcntl(fd, F_SETFD, rval | FD_CLOEXEC | O_NONBLOCK)) != -1, perror("fcntl(F_SETFD)"));
+#else
+    VerifyOrExit((fd = socket(aDomain, aType | SOCK_CLOEXEC | SOCK_NONBLOCK, aProtocol)) != -1,
+                 perror("socket(SOCK_CLOEXEC)"));
+#endif
+
+exit:
+    if (rval == -1 && fd != -1)
```
Comment (simonlingoogle) [inline src/posix/platform/misc.cpp:126]: nit: `fd` must not be -1 if `rval == -1` ?
```suggestion
    if (rval == -1)
```

File: src/posix/platform/platform-posix.h [hunk 2]
```diff
@@ -442,6 +442,20 @@ void platformUdpUpdateFdSet(otInstance *aInstance, fd_set *aReadFdSet, int *aMax
...
+ * @param[in]   aDomain     The communication domain.
+ * @param[in]   aType       The semantics of communication.
+ * @param[in]   aProtocol   The protocol to use.
+ *
+ * @returns The file descriptor of the created socket.
+ *
+ * @retval  -1  Failed to create socket.
+ *
+ */
+int NonBlockSocketWithCloseExec(int aDomain, int aType, int aProtocol);
```
Comment (bukepo) [inline src/posix/platform/platform-posix.h:457]: Is it possible to merge into SocketWithCloseExec() by adding an additional argument for nonblocking?

File: src/posix/platform/misc.cpp [hunk 3]
```diff
@@ -92,13 +92,21 @@ int SocketWithCloseExec(int aDomain, int aType, int aProtocol)
...
+    if (aNonBlock)
+    {
+        rval |= O_NONBLOCK;
+    }
+    VerifyOrExit((rval = fcntl(fd, F_SETFD, rval)) != -1, perror("fcntl(F_SETFD)"));
 #else
-    VerifyOrExit((fd = socket(aDomain, aType | SOCK_CLOEXEC, aProtocol)) != -1, perror("socket(SOCK_CLOEXEC)"));
+    int type;
+
+    type = aNonBlock ? aType | SOCK_CLOEXEC | SOCK_NONBLOCK : aType | SOCK_CLOEXEC;
```
Comment (bukepo) [inline src/posix/platform/misc.cpp:104]: use `aType` instead of defining a new variable?

File: src/posix/platform/misc.cpp [hunk 4]
```diff
@@ -92,13 +92,19 @@ int SocketWithCloseExec(int aDomain, int aType, int aProtocol)
     VerifyOrExit((fd = socket(aDomain, aType, aProtocol)) != -1, perror("socket(SOCK_CLOEXEC)"));
 
     VerifyOrExit((rval = fcntl(fd, F_GETFD, 0)) != -1, perror("fcntl(F_GETFD)"));
-    VerifyOrExit((rval = fcntl(fd, F_SETFD, rval | FD_CLOEXEC)) != -1, perror("fcntl(F_SETFD)"));
+    rval |= FD_CLOEXEC;
+    if (aNonBlock)
+    {
+        rval |= O_NONBLOCK;
+    }
```
Comment (simonlingoogle) [inline src/posix/platform/misc.cpp:99]:
```suggestion
    rval |= aNonBlock ? O_NONBLOCK | FD_CLOEXEC : FD_CLOEXEC;
```
Why are we using `?:` operator in `#else` block but not here?

File: src/posix/platform/misc.cpp [hunk 5]
```diff
@@ -92,13 +92,19 @@ int SocketWithCloseExec(int aDomain, int aType, int aProtocol)
...
+    VerifyOrExit((rval = fcntl(fd, F_SETFD, rval)) != -1, perror("fcntl(F_SETFD)"));
 #else
-    VerifyOrExit((fd = socket(aDomain, aType | SOCK_CLOEXEC, aProtocol)) != -1, perror("socket(SOCK_CLOEXEC)"));
+    aType = aNonBlock ? aType | SOCK_CLOEXEC | SOCK_NONBLOCK : aType | SOCK_CLOEXEC;
```
Comment (simonlingoogle) [inline src/posix/platform/misc.cpp:102]:
```suggestion
    aType |= aNonBlock ? SOCK_CLOEXEC | SOCK_NONBLOCK : SOCK_CLOEXEC;
```

File: src/posix/platform/platform-posix.h [hunk 6]
```diff
@@ -434,13 +434,14 @@ void platformUdpUpdateFdSet(otInstance *aInstance, fd_set *aReadFdSet, int *aMax
...
  * @param[in]   aProtocol   The protocol to use.
+ * @param[in]   aNonBlock   Whether to add nonblock flags.
  *
  * @returns The file descriptor of the created socket.
  *
  * @retval  -1  Failed to create socket.
  *
  */
-int SocketWithCloseExec(int aDomain, int aType, int aProtocol);
+int SocketWithCloseExec(int aDomain, int aType, int aProtocol, bool aNonBlock);
```
[the remark this answers is not in the document]
Comment (gjc13) [reply src/posix/platform/platform-posix.h:444]: Not sure whether it's a good idea @bukepo @jwhui . Is it common practice to add such definitions in OpenThread?

[the remark this answers is not in the document]
Comment (jwhui) [reply src/posix/platform/platform-posix.h:444]: Named values are generally preferred for the reasons that @rob-the-dude mentioned. Some examples in existing code:

https://github.com/openthread/openthread/blob/af4837d122c9da0eb6dcb9f2a1b19b1e37b738cc/src/core/mac/data_poll_sender.hpp#L274-L278

https://github.com/openthread/openthread/blob/af4837d122c9da0eb6dcb9f2a1b19b1e37b738cc/src/core/mac/mac_types.hpp#L111-L115

https://github.com/openthread/openthread/blob/af4837d122c9da0eb6dcb9f2a1b19b1e37b738cc/src/core/thread/mle.hpp#L1049-L1053

[the remark this answers is not in the document]
Comment (gjc13) [reply src/posix/platform/platform-posix.h]: Done

**Pull request merged**